tasks/main.yml

```diff
@@ -223,15 +225,65 @@
         state: absent
       when: __mssql_curr_repo.stdout | length > 0
 
+- name: Prepare for the upgrade
+  when:
+    - __mssql_current_version is defined
+    - mssql_upgrade | bool
+    - __mssql_current_version is defined
+    - ansible_pkg_mgr == "zypper"
+  block:
+    # This works only on systems that use yum or dnf
+    - name: >-
+        Update all packages from SQL Server repo
+        version {{ __mssql_current_version }}
+      package:
+        name: "*"
+        extra_args: >-
+          -r packages-microsoft-com-mssql-server-{{ __mssql_current_version }}
+        state: latest  # noqa package-latest
+
+    - name: Get zypper repositories  # noqa command-instead-of-module
+      shell: >-
+        set -euo pipefail;
+        zypper repos -E
+        | grep packages-microsoft-com-mssql-server-
+        | grep -v packages-microsoft-com-mssql-server-{{ mssql_version }}
+        | cut -d" " -f 4
+        || true
+      register: __mssql_curr_repo
+      changed_when: false
+      check_mode: false
+
+    - name: >-
+        Remove the current Microsoft SQL Server repository to upgrade to
+        {{ mssql_version }}
+      community.general.zypper_repository:
```

Comment:
Is there no way to do this with the package module?  The problem with using third party modules like this is that the module is evaluated at **file load** time - which means **all** users on **all** platforms will need to have installed `community.general` or **the role will fail**.  So why doesn't the `when` clause prevent this?  Because `when` is evaluated at **file processing** time, which is after **file load** time.

If there is no way to do this with the `package` module or other Ansible built-in, then maybe you could put all of the zypper stuff in a file `tasks/suse_packaging.yml` or something like that, then in `tasks/main.yml` do something like
```yaml
- name: Handle Suse packaging
  include_tasks: suse_packaging.yml
  when: ansible_facts["distribution"] == "Suse"
```
In this case, the `include_tasks` and `when` will be processed at **file processing** time, which means only Suse users will need to install the 3rd party collection.

Reply:
Or maybe you can safely do this using `command` or `shell` if this is possible.

Reply:
I have changed it to use zypper commands directly